src/components/services/WalletActivator.js

```diff
@@ -0,0 +1,129 @@
+type Subscriptions = { [walletId: string]: (() => mixed) | void }
+
+export function WalletActivatorComponent(props: Props) {
+  const { account, context, guiWallets } = props
+
+  // Subscribe to the context paused flag:
+  const [paused, setPaused] = useState(false)
+  useEffect(() => {
+    if (typeof context.watch !== 'function') return
+    const cleanup = context.watch('paused', setPaused)
+    return () => {
+      cleanup()
+    }
+  })
+
+  // Subscribe to the core wallet list:
+  const [activeWalletIds, setActiveWalletIds] = useState(account.activeWalletIds)
+  const [currencyWallets, setCurrencyWallets] = useState(account.currencyWallets)
+  useEffect(() => {
+    if (typeof account.watch !== 'function') return
+
+    const cleanups = [account.watch('activeWalletIds', setActiveWalletIds), account.watch('currencyWallets', setCurrencyWallets)]
+    return () => {
+      for (const cleanup of cleanups) cleanup()
+    }
+  }, [account])
+
+  // Wallets we have finished, or have timed out:
+  const timers: { +current: Timers } = useRef({})
+  const subscriptions: { +current: Subscriptions } = useRef({})
+  const [completeWallets, setCompleteWallets] = useState(emptyWalletIds)
+
+  // Activate wallets we need:
+  useEffect(() => {
+    if (typeof account.watch !== 'function') return
+
+    // If we are offline, turn off all the wallets and clear our state:
+    if (paused) {
+      if (completeWallets !== emptyWalletIds) {
+        setCompleteWallets(emptyWalletIds)
+      }
+      for (const id of Object.keys(timers.current)) {
+        clearTimeout(timers.current[id])
+        timers.current[id] = undefined
+      }
+      for (const id of Object.keys(subscriptions.current)) {
+        const cleanup = subscriptions.current[id]
+        if (cleanup != null) cleanup()
+        subscriptions.current[id] = undefined
+      }
+      for (const walletId of account.activeWalletIds) {
+        const wallet = currencyWallets[walletId]
+        if (wallet == null || wallet.paused) continue
+        wallet.changePaused(true).catch(showError)
+      }
+      return
+    }
+
+    // We are online, so make sure at least 3 wallets are running:
+    let runningWallets = 0
+    for (const walletId of account.activeWalletIds) {
+      // Skip wallets that haven't loaded yet, or have already finished:
+      if (currencyWallets[walletId] == null || guiWallets[walletId] == null || completeWallets.includes(walletId)) {
+        continue
+      }
+
+      // If we already have 3 wallets starting up, we are done:
+      if (++runningWallets >= 3) break
+
+      // If the wallet is already started, we don't need to do anything more:
+      if (!currencyWallets[walletId].paused) continue
+
+      // Start the wallet:
+      currencyWallets[walletId].changePaused(false).catch(showError)
+
+      // Set a timeout:
+      timers.current[walletId] = setTimeout(() => {
+        timers.current[walletId] = undefined
```

Comment:
It might be safer (also prettier) to do it like this:

```
const { [walletId]: _, ...rest } = timers.current 
timers.current = rest
```

Or even call the "cleanTimer" function from below, which would probably require adding the following validation to the funcion:
`if (timer === null) return`